subsys/bluetooth/host/crypto.c

```diff
@@ -114,6 +119,35 @@ int bt_rand(void *buf, size_t len)
 
 	return -EIO;
 }
+#else /* !CONFIG_BT_HOST_CRYPTO_RANDOM */
+int bt_rand(void *buf, size_t len)
+{
+	int ret, size;
+	size_t i = 0;
+
+	while (len) {
+		struct bt_hci_rp_le_rand *rp;
+		struct net_buf *rsp;
+
+		ret = bt_hci_cmd_send_sync(BT_HCI_OP_LE_RAND, NULL, &rsp);
+		if (ret) {
+			return ret;
+		}
```

Comment:
@Thalley We use controller (https://github.com/zephyrproject-rtos/zephyr/pull/42264#discussion_r814699937)

Reply:
This scheme is to directly call the BT Rand HCI command, while previously, by using the BT Rand HCI command as a random seed, both will call the controller.

Reply:
Right, but why are doing the check for supporting `HCI_LE_Rand` in a function that doesn't call `bt_hci_cmd_send_sync(BT_HCI_OP_LE_RAND, NULL, &rsp);`, rather than doing it here? 

What I suggest is to guard the entire `prng_init` function with `CONFIG_BT_HOST_CRYPTO_RANDOM` and then add the 
```c
if (!BT_CMD_TEST(bt_dev.supported_commands, 27, 7)) {
		return -ENOTSUP;
}
```
check here as well. 

Using `prng_init` just for a controller check seem like a poor design to me.

Reply:
It's not friendly in my opinion to have to make judgments before each call. During initialization. Judging once is the easiest way.

Reply:
Judging before each call is how most other HCI calls are implemented. 

The current implementation of `bt_rand` when `CONFIG_BT_HOST_CRYPTO_RANDOM=y` tests for `if (ret == TC_HMAC_PRNG_RESEED_REQ)` which is implicitly a check for `BT_CMD_TEST(bt_dev.supported_commands, 27, 7)`, so we already check each time now. 

Furthermore, the only place where `prng_init` is called is in `common_init` where the call is guarded by `IS_ENABLED(CONFIG_BT_HOST_CRYPTO)`, so there's absolutely no reason not to guard `prng_init` with CONFIG_BT_HOST_CRYPTO. For that same reason, the check for `BT_CMD_TEST(bt_dev.supported_commands, 27, 7)` in `prng_init` is, in this PRs current state, never done when `CONFIG_BT_HOST_CRYPTO_RANDOM=n`

Reply:
CC?

Reply:
I want to try to avoid the situation that the controller does not support bt rand, but the host initiates this command.

Reply:
Anyway, I have made changes, but I still want to state that BT_HOST_CRYPTO_RANDOM depends on BT_HOST_CRYPTO.

Even if BT_HOST_CRYPTO_RANDOM=y or =n, it still makes sense for the initialization function prng_init to exist.

Reply:
> if (ret == TC_HMAC_PRNG_RESEED_REQ) which is implicitly a check for BT_CMD_TEST(bt_dev.supported_commands, 27, 7), so we already check each time now.

Still trying to understand is there aimplicitly a check here, I don't understand.

`TC_HMAC_PRNG_RESEED_REQ` is tinycrypt, nothing to do with bluetooth.

Reply:
> > if (ret == TC_HMAC_PRNG_RESEED_REQ) which is implicitly a check for BT_CMD_TEST(bt_dev.supported_commands, 27, 7), so we already check each time now.
> 
> Still trying to understand is there implicitly a check here, I don't understand.
> 
> `TC_HMAC_PRNG_RESEED_REQ` is tinycrypt, nothing to do with bluetooth.

The host implementation of `bt_rand` calls `tc_hmac_prng_generate` and handles `TC_HMAC_PRNG_RESEED_REQ` so it is related to Bluetooth. 

I still stand by that calling `prng_init` only as a check for `if (!BT_CMD_TEST(bt_dev.supported_commands, 27, 7))` before calling calling `bt_rand`, instead of moving the check for `if (!BT_CMD_TEST(bt_dev.supported_commands, 27, 7))` into `bt_rand` is a poor design. 

Not only is `prng_init` not a public function (and thus the application can never get that result), but it is a function that is entirely unrelated to calling `bt_rand` when using the HCI command to generate a random value. 

In the case that an application tries to call `bt_rand`, and the controller does not support it, this will just result in a `-EIO` return code, which will be difficult for the application to debug. 

So I still suggest guarding `prng_init` with `CONFIG_BT_HOST_CRYPTO_RANDOM` and moving the check for controller support to `bt_rand` if `CONFIG_BT_HOST_CRYPTO_RANDOM=n`

Reply:
This time it seems to convince me.

Reply:
My apologies for being unclear :)